proof/refine/InvariantUpdates_H.thy

```diff
@@ -0,0 +1,457 @@
+qed
+
+(* FIXME arch-split: export? *)
+lemma invs_no_cicd'_machine:
+  assumes mask: "irq_masks (f (ksMachineState s)) =
+                 irq_masks (ksMachineState s)"
+  assumes vms: "valid_machine_state' (ksMachineState_update f s) =
+                valid_machine_state' s"
+  shows "invs_no_cicd' (ksMachineState_update f s) = invs_no_cicd' s"
+proof -
+  show ?thesis
+    apply (cases "ksSchedulerAction s")
+    apply (simp_all add: all_invs_but_ct_idle_or_in_cur_domain'_def valid_state'_def
+                         cur_tcb'_def ct_in_state'_def ct_idle_or_in_cur_domain'_def
+                         tcb_in_cur_domain'_def valid_bitmaps_def bitmapQ_defs
+                         vms ct_not_inQ_def
+                         state_refs_of'_def ps_clear_def
+                         valid_irq_node'_def mask
+                   cong: option.case_cong)
+    done
+qed
+
+lemma pspace_no_overlap_queues [simp]:
+  "pspace_no_overlap' w sz (ksReadyQueues_update f s) = pspace_no_overlap' w sz s"
+  by (simp add: pspace_no_overlap'_def)
+
+lemma pspace_no_overlap'_ksSchedulerAction[simp]:
+  "pspace_no_overlap' a b (ksSchedulerAction_update f s) =
+   pspace_no_overlap' a b s"
+  by (simp add: pspace_no_overlap'_def)
+
+lemma pspace_no_overlap'_ksArchState_update[simp]:
+  "pspace_no_overlap' p n (ksArchState_update f s) =
+   pspace_no_overlap' p n s"
+  by (simp add: pspace_no_overlap'_def)
+
+lemma ksReadyQueues_update_id[simp]:
+  "ksReadyQueues_update id s = s"
+  by simp
+
+lemma ct_not_inQ_ksReadyQueues_update[simp]:
+  "ct_not_inQ (ksReadyQueues_update f s) = ct_not_inQ s"
+  by (simp add: ct_not_inQ_def)
+
+lemma inQ_context[simp]:
+  "inQ d p (tcbArch_update f tcb) = inQ d p tcb"
+  by (cases tcb, simp add: inQ_def)
+
+lemma valid_tcb'_tcbQueued[simp]:
+  "valid_tcb' (tcbQueued_update f tcb) = valid_tcb' tcb"
+  by (cases tcb, rule ext, simp add: valid_tcb'_def tcb_cte_cases_def tcb_cte_cases_neqs)
+
+lemma valid_tcb'_tcbFault_update[simp]:
+  "valid_tcb' tcb s \<Longrightarrow> valid_tcb' (tcbFault_update f tcb) s"
+  by (clarsimp simp: valid_tcb'_def tcb_cte_cases_def tcb_cte_cases_neqs)
+
+lemma valid_tcb'_tcbTimeSlice_update[simp]:
+  "valid_tcb' (tcbTimeSlice_update f tcb) s = valid_tcb' tcb s"
+  by (simp add: valid_tcb'_def tcb_cte_cases_def tcb_cte_cases_neqs)
+
+lemma invs'_wu[simp]:
+  "invs' (ksWorkUnitsCompleted_update f s) = invs' s"
+  by (simp add: invs'_def cur_tcb'_def valid_state'_def valid_bitmaps_def
+                valid_irq_node'_def valid_machine_state'_def
+                ct_not_inQ_def ct_idle_or_in_cur_domain'_def tcb_in_cur_domain'_def
+                bitmapQ_defs)
+
+lemma if_unsafe_then_cap_arch'[simp]:
+  "if_unsafe_then_cap' (ksArchState_update f s) = if_unsafe_then_cap' s"
+  by (simp add: if_unsafe_then_cap'_def ex_cte_cap_to'_def)
+
+lemma valid_bitmaps_ksSchedulerAction_update[simp]:
+   "valid_bitmaps (ksSchedulerAction_update f s) = valid_bitmaps s"
+   by (simp add: valid_bitmaps_def bitmapQ_defs)
```

Comment:
Do we care about fixing style while moving these lemmas? If we do, this is overindented.

Reply:
Fixed, thanks.